torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1341,13 +1385,21 @@ void ProcessGroupNCCL::heartbeatMonitor() {
   // Case two: desync might be slow or get stuck. Or we get stuck in
   // destructors, we will sleep for some time before calling std::abort() to
   // kill the whole process.
-  if ((terminateProcessGroup_.load() || collectiveDebugInfoMode_.load()) &&
+  if ((terminateProcessGroup_.load() || collectiveDebugInfoMode_.load() ||
+       !debugInfoStoreThread.joinable()) &&
       !terminateHeartbeatMonitorThread_.load()) {
     // Leave another two mins for desync report generation or process group
     // destroy.
     std::this_thread::sleep_for(std::chrono::seconds(heartbeatTimeoutInSec_));
   }
 
+  // At this point, we either already sleep for another `heartbeatTimeoutInSec_`
+  // or the thread has finished. Because we don't want to block the monitor
+  // thread, so We mark the thread detach and the dump of debug info becomes
+  // "best effort". If the process exit normally, marking it detach also makes
```

Comment:
if the process exits normally, is it possible for the heartbeat monitor to reach this line? if we ever exit the loop above, does it mean an error happened?

Reply:
If the process exits normally, we will early return, then no need to dump the debug info right? And yes, it's still possible, but you are right, I need to add a `joinable()` check, I was under the wrong impression that we can detach a default constructed thread..